{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ignasi/Desktop/kagle/models/DRfold2/cfg_97/EvoMSA2XYZ.py:28: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.\n",
      "  RNAlm.load_state_dict(torch.load(saved_model,map_location=torch.device('cpu')),strict=False)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_273991/3146032445.py:44: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.\n",
      "  RNAlm.load_state_dict(torch.load(saved_model,map_location=torch.device('cpu')),strict=False)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "RNA2nd(\n",
       "  (emb): RNAembedding(\n",
       "    (qlinear): Linear(\n",
       "      (linear): Linear(in_features=6, out_features=128, bias=True)\n",
       "    )\n",
       "    (klinear): Linear(\n",
       "      (linear): Linear(in_features=6, out_features=128, bias=True)\n",
       "    )\n",
       "    (slinear): Linear(\n",
       "      (linear): Linear(in_features=6, out_features=512, bias=True)\n",
       "    )\n",
       "    (zlinear): Linear(\n",
       "      (linear): Linear(in_features=3, out_features=128, bias=True)\n",
       "    )\n",
       "    (poslinears): Linear(\n",
       "      (linear): Linear(in_features=64, out_features=512, bias=True)\n",
       "    )\n",
       "    (poslinearz): Linear(\n",
       "      (linear): Linear(in_features=64, out_features=128, bias=True)\n",
       "    )\n",
       "  )\n",
       "  (evmodel): Evoformer(\n",
       "    (evos): ModuleList(\n",
       "      (0-17): 18 x EvoBlock(\n",
       "        (msa_row): MSARow(\n",
       "          (norm1): LayerNorm((512,), eps=1e-05, elementwise_affine=True)\n",
       "          (qlinear): LinearNoBias(\n",
       "            (linear): Linear(in_features=512, out_features=256, bias=False)\n",
       "          )\n",
       "          (klinear): LinearNoBias(\n",
       "            (linear): Linear(in_features=512, out_features=256, bias=False)\n",
       "          )\n",
       "          (vlinear): LinearNoBias(\n",
       "            (linear): Linear(in_features=512, out_features=256, bias=False)\n",
       "          )\n",
       "          (norm_z): LayerNorm((128,), eps=1e-05, elementwise_affine=True)\n",
       "          (zlinear): LinearNoBias(\n",
       "            (linear): Linear(in_features=128, out_features=16, bias=False)\n",
       "          )\n",
       "          (glinear): Linear(\n",
       "            (linear): Linear(in_features=512, out_features=256, bias=True)\n",
       "          )\n",
       "          (olinear): Linear(\n",
       "            (linear): Linear(in_features=256, out_features=512, bias=True)\n",
       "          )\n",
       "        )\n",
       "        (msa_col): MSACol(\n",
       "          (norm1): LayerNorm((512,), eps=1e-05, elementwise_affine=True)\n",
       "          (qlinear): LinearNoBias(\n",
       "            (linear): Linear(in_features=512, out_features=256, bias=False)\n",
       "          )\n",
       "          (klinear): LinearNoBias(\n",
       "            (linear): Linear(in_features=512, out_features=256, bias=False)\n",
       "          )\n",
       "          (vlinear): LinearNoBias(\n",
       "            (linear): Linear(in_features=512, out_features=256, bias=False)\n",
       "          )\n",
       "          (glinear): Linear(\n",
       "            (linear): Linear(in_features=512, out_features=256, bias=True)\n",
       "          )\n",
       "          (olinear): Linear(\n",
       "            (linear): Linear(in_features=256, out_features=512, bias=True)\n",
       "          )\n",
       "        )\n",
       "        (msa_trans): MSATrans(\n",
       "          (norm): LayerNorm((512,), eps=1e-05, elementwise_affine=True)\n",
       "          (linear1): Linear(\n",
       "            (linear): Linear(in_features=512, out_features=1024, bias=True)\n",
       "          )\n",
       "          (linear2): Linear(\n",
       "            (linear): Linear(in_features=1024, out_features=512, bias=True)\n",
       "          )\n",
       "        )\n",
       "        (msa_opm): MSAOPM(\n",
       "          (norm): LayerNorm((512,), eps=1e-05, elementwise_affine=True)\n",
       "          (linear1): Linear(\n",
       "            (linear): Linear(in_features=512, out_features=12, bias=True)\n",
       "          )\n",
       "          (linear2): Linear(\n",
       "            (linear): Linear(in_features=512, out_features=12, bias=True)\n",
       "          )\n",
       "          (linear3): Linear(\n",
       "            (linear): Linear(in_features=144, out_features=128, bias=True)\n",
       "          )\n",
       "        )\n",
       "        (pair_triout): TriOut(\n",
       "          (norm): LayerNorm((128,), eps=1e-05, elementwise_affine=True)\n",
       "          (onorm): LayerNorm((72,), eps=1e-05, elementwise_affine=True)\n",
       "          (alinear): Linear(\n",
       "            (linear): Linear(in_features=128, out_features=72, bias=True)\n",
       "          )\n",
       "          (blinear): Linear(\n",
       "            (linear): Linear(in_features=128, out_features=72, bias=True)\n",
       "          )\n",
       "          (aglinear): Linear(\n",
       "            (linear): Linear(in_features=128, out_features=72, bias=True)\n",
       "          )\n",
       "          (bglinear): Linear(\n",
       "            (linear): Linear(in_features=128, out_features=72, bias=True)\n",
       "          )\n",
       "          (glinear): Linear(\n",
       "            (linear): Linear(in_features=128, out_features=128, bias=True)\n",
       "          )\n",
       "          (olinear): Linear(\n",
       "            (linear): Linear(in_features=72, out_features=128, bias=True)\n",
       "          )\n",
       "        )\n",
       "        (pair_triin): TriIn(\n",
       "          (norm): LayerNorm((128,), eps=1e-05, elementwise_affine=True)\n",
       "          (onorm): LayerNorm((72,), eps=1e-05, elementwise_affine=True)\n",
       "          (alinear): Linear(\n",
       "            (linear): Linear(in_features=128, out_features=72, bias=True)\n",
       "          )\n",
       "          (blinear): Linear(\n",
       "            (linear): Linear(in_features=128, out_features=72, bias=True)\n",
       "          )\n",
       "          (aglinear): Linear(\n",
       "            (linear): Linear(in_features=128, out_features=72, bias=True)\n",
       "          )\n",
       "          (bglinear): Linear(\n",
       "            (linear): Linear(in_features=128, out_features=72, bias=True)\n",
       "          )\n",
       "          (glinear): Linear(\n",
       "            (linear): Linear(in_features=128, out_features=128, bias=True)\n",
       "          )\n",
       "          (olinear): Linear(\n",
       "            (linear): Linear(in_features=72, out_features=128, bias=True)\n",
       "          )\n",
       "        )\n",
       "        (pair_tristart): TriAttStart(\n",
       "          (norm): LayerNorm((128,), eps=1e-05, elementwise_affine=True)\n",
       "          (qlinear): Linear(\n",
       "            (linear): Linear(in_features=128, out_features=32, bias=True)\n",
       "          )\n",
       "          (klinear): Linear(\n",
       "            (linear): Linear(in_features=128, out_features=32, bias=True)\n",
       "          )\n",
       "          (vlinear): Linear(\n",
       "            (linear): Linear(in_features=128, out_features=32, bias=True)\n",
       "          )\n",
       "          (blinear): Linear(\n",
       "            (linear): Linear(in_features=128, out_features=4, bias=True)\n",
       "          )\n",
       "          (glinear): Linear(\n",
       "            (linear): Linear(in_features=128, out_features=32, bias=True)\n",
       "          )\n",
       "          (olinear): Linear(\n",
       "            (linear): Linear(in_features=32, out_features=128, bias=True)\n",
       "          )\n",
       "        )\n",
       "        (pair_triend): TriAttEnd(\n",
       "          (norm): LayerNorm((128,), eps=1e-05, elementwise_affine=True)\n",
       "          (qlinear): Linear(\n",
       "            (linear): Linear(in_features=128, out_features=32, bias=True)\n",
       "          )\n",
       "          (klinear): Linear(\n",
       "            (linear): Linear(in_features=128, out_features=32, bias=True)\n",
       "          )\n",
       "          (vlinear): Linear(\n",
       "            (linear): Linear(in_features=128, out_features=32, bias=True)\n",
       "          )\n",
       "          (blinear): Linear(\n",
       "            (linear): Linear(in_features=128, out_features=4, bias=True)\n",
       "          )\n",
       "          (glinear): Linear(\n",
       "            (linear): Linear(in_features=128, out_features=32, bias=True)\n",
       "          )\n",
       "          (olinear): Linear(\n",
       "            (linear): Linear(in_features=32, out_features=128, bias=True)\n",
       "          )\n",
       "        )\n",
       "        (pair_trans): PairTrans(\n",
       "          (norm): LayerNorm((128,), eps=1e-05, elementwise_affine=True)\n",
       "          (linear1): Linear(\n",
       "            (linear): Linear(in_features=128, out_features=256, bias=True)\n",
       "          )\n",
       "          (linear2): Linear(\n",
       "            (linear): Linear(in_features=256, out_features=128, bias=True)\n",
       "          )\n",
       "        )\n",
       "      )\n",
       "    )\n",
       "  )\n",
       "  (seq_head): Linear(\n",
       "    (linear): Linear(in_features=512, out_features=5, bias=True)\n",
       "  )\n",
       "  (joint_head): Linear(\n",
       "    (linear): Linear(in_features=128, out_features=25, bias=True)\n",
       "  )\n",
       ")"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sys\n",
    "import os\n",
    "\n",
    "expdir = os.getcwd()\n",
    "sys.path.append(os.path.join(expdir,'cfg_97'))\n",
    "\n",
    "import numpy as np\n",
    "from cfg_97 import data\n",
    "import torch\n",
    "import torch.nn.functional as F\n",
    "import os\n",
    "import sys \n",
    "from cfg_97.EvoMSA2XYZ import MSA2XYZ\n",
    "from cfg_97.RNALM2 import Model\n",
    "\n",
    "DEVICE = torch.device('cuda:1')  # For GPU 1\n",
    "basenpy_standard= np.load(os.path.join(expdir,'cfg_97','base.npy'  )  )\n",
    "def data_collect(pdb_seq):\n",
    "    aa_type = data.parse_seq(pdb_seq)\n",
    "    base = data.Get_base(pdb_seq,basenpy_standard)\n",
    "    seq_idx = np.arange(len(pdb_seq)) + 1\n",
    "\n",
    "    msa=aa_type[None,:]\n",
    "    msa=torch.from_numpy(msa).to(DEVICE)\n",
    "    msa=torch.cat([msa,msa],0)\n",
    "    msa=F.one_hot(msa.long(),6).float()\n",
    "\n",
    "    base_x = torch.from_numpy(base).float().to(DEVICE)\n",
    "    seq_idx = torch.from_numpy(seq_idx).long().to(DEVICE)\n",
    "    return msa,base_x,seq_idx\n",
    "\n",
    "\n",
    "#RCLM\n",
    "lmcfg={}\n",
    "lmcfg['s_in_dim']=5\n",
    "lmcfg['z_in_dim']=2\n",
    "lmcfg['s_dim']= 512\n",
    "lmcfg['z_dim']= 128\n",
    "lmcfg['N_elayers']=18\n",
    "RNAlm = Model.RNA2nd(lmcfg)\n",
    "\n",
    "#Load RCLM\n",
    "saved_model =  os.path.join(os.getcwd(), 'model_hub','RCLM', 'epoch_67000')\n",
    "RNAlm.load_state_dict(torch.load(saved_model,map_location=torch.device('cpu')),strict=False)\n",
    "RNAlm.to(DEVICE)\n",
    "RNAlm.eval()\n",
    "\n",
    "# msa N L D, seq L D\n",
    "# mask: maksing, L, 1 means masked\n",
    "# aa:   L x s_in_dim\n",
    "# ss:   L x L x 2\n",
    "# idx:  L (LongTensor)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "path_to_test_data = \"../../data/\"\n",
    "output_directory = \"fasta_files\"\n",
    "test_data = pd.read_csv(path_to_test_data+\"test_sequences.csv\")\n",
    "\n",
    "def generate_fasta_file(test_data):\n",
    "    sequences = test_data[\"sequence\"].tolist()\n",
    "    target_id = test_data[\"target_id\"].tolist()\n",
    "    for seq, target in zip(sequences, target_id):\n",
    "        with open(f\"{output_directory}/{target}.fasta\", \"a\") as f:\n",
    "            f.write(f\">{target}\\n{seq}\\n\")\n",
    "            \n",
    "generate_fasta_file(test_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MSA size:  torch.Size([2, 369, 6])\n",
      "Base size:  torch.Size([369, 3, 3])\n",
      "Seq idx size:  torch.Size([369])\n"
     ]
    }
   ],
   "source": [
    "fasta_files = os.listdir(output_directory)\n",
    "fasta_file = \"R1107.fasta\"\n",
    "infasta = os.path.join(output_directory, fasta_file)\n",
    "lines = open(infasta).readlines()[1:]\n",
    "seqs = [aline.strip() for aline in lines]\n",
    "seq = ''.join(seqs)\n",
    "\n",
    "def data_collect_seq(pdb_seq):\n",
    "    aa_type = data.parse_seq(pdb_seq)\n",
    "    base = data.Get_base(pdb_seq,basenpy_standard)\n",
    "    seq_idx = np.arange(len(pdb_seq)) + 1\n",
    "\n",
    "    msa=aa_type[None,:]\n",
    "    msa=torch.from_numpy(msa).to(DEVICE)\n",
    "    msa=torch.cat([msa,msa],0)\n",
    "    msa=F.one_hot(msa.long(),6).float()\n",
    "\n",
    "    base_x = torch.from_numpy(base).float().to(DEVICE)\n",
    "    seq_idx = torch.from_numpy(seq_idx).long().to(DEVICE)\n",
    "    return msa,base_x,seq_idx\n",
    "    \n",
    "#This returns msa, base_x, and seq_idx in device already\n",
    "msa,base_x,seq_idx = data_collect_seq(seq)\n",
    "\n",
    "print(\"MSA size: \", msa.size())\n",
    "print(\"Base size: \", base_x.size())\n",
    "print(\"Seq idx size: \", seq_idx.size())\n",
    "\n",
    "#Not sure why MSA is concatenated twice maybe for pair representation?\n",
    "#Base size contains template 3x3 coordinates for [P, C4', N1/N09]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## We will run just one iteration of MSA2xyz to understand what are the inputs/outputs.\n",
    "\n",
    "### MSA2xyz iteration generates a Pair Matrix and MSA features and those are given to the Structure Module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n",
      "will do checkpoint\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ignasi/anaconda3/envs/pytorch_venv/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py:632: UserWarning: torch.utils.checkpoint: the use_reentrant parameter should be passed explicitly. In version 2.5 we will raise an exception if use_reentrant is not passed. use_reentrant=False is recommended, but if you need to preserve the current default behavior, you can pass use_reentrant=True. Refer to docs for more details on the differences between the two variants.\n",
      "  return fn(*args, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "from cfg_97 import EvoMSA2XYZ\n",
    "from cfg_97.EvoMSA2XYZ import MSA2xyzIteration\n",
    "\n",
    "# one-hot encoded dims - [A, U, G, C, gap (-), mask, padding]\n",
    "msa_dim= 7\n",
    "# one-hot encoded dims - [A, U, G, C, gap (-), mask]\n",
    "seq_dim = 6\n",
    "#embeddings dims\n",
    "m_dim,s_dim,z_dim = 64,64,64\n",
    "#N ensembles and N cycles\n",
    "N_ensemble,N_cycle=1,1\n",
    "#Length of the sequence\n",
    "L = msa.size(1)\n",
    "\n",
    "#Previous distance: pairwise distances between nucleotides - OUTPUT of the structure module (from one cycle)\n",
    "## [0-2Å], [2-2.5Å], [2.5-3Å], ..., [>20Å]\n",
    "previous_dis = torch.zeros([L,L,38]).to(msa.device) \n",
    "previous_dis[...,0] = 1\n",
    "\n",
    "#Previous helix-bond - OUTPUT of the structure module (from one cycle)\n",
    "#Types of helix bonds: A-U pair, G-C pair, Wobble G-U, Non-canonical, Stacking, no interaction?\n",
    "previous_hb = torch.zeros([L,L,6]).to(msa.device)\n",
    "previous_hb[...,0] = 1\n",
    "\n",
    "m1_pre,z_pre=0,0\n",
    "x_pre=torch.zeros(L,3,3).to(msa.device)\n",
    "index = 0 \n",
    "\n",
    "model = MSA2xyzIteration(seq_dim,msa_dim,N_ensemble,m_dim,s_dim,z_dim)\n",
    "model = model.to(DEVICE)\n",
    "model.eval()\n",
    "\n",
    "# Move model to the specified device\n",
    "alphas = np.array(list(seq))\n",
    "previous_dis = previous_dis.to(DEVICE) \n",
    "previous_hb = previous_hb.to(DEVICE)\n",
    "\n",
    "# Now make prediction\n",
    "preds = model.pred(msa, seq_idx, None, m1_pre, z_pre, x_pre, index, alphas, previous_dis, previous_hb)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "torch.Size([369, 64]) torch.Size([369, 369, 64]) torch.Size([369, 64])\n"
     ]
    }
   ],
   "source": [
    "m, z, s = preds\n",
    "print(m.shape, z.shape, s.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #Load MSA2XYZ\n",
    "# model = MSA2XYZ(seq_dim=21,msa_dim=6,N_ensemble=1,N_cycle=1,m_dim=64,s_dim=128,z_dim=64,docheck=True)\n",
    "\n",
    "# #run structure model to get scoring\n",
    "# #run optimization to get final structure\n",
    "# #can we optimize without running structure model?\n",
    "# #ccan we direcly get final pdf?\n",
    "# #run arena to get final pdb\n",
    "\n",
    "# #run moddel to get coordinates\n",
    "# def run_model(model, fasta_file, prefix):\n",
    "#     with torch.no_grad():\n",
    "#         lines = open(fasta_file).readlines()[1:]\n",
    "#         seqs = [aline.strip() for aline in lines]\n",
    "#         seq = ''.join(seqs)\n",
    "#         msa,base_x,seq_idx = data_collect(seq)\n",
    "#         ret = model.pred(msa,seq_idx,None,base_x,np.array(list(seq)))\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pytorch_venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
